### PR TITLE
7903722: JMH: Add xctrace-based perfnorm profiler for macOS

### PR DESCRIPTION
**Pull request by fzhinkin**
Implementation of a perfnorm-alike profiler for macOS based on `xctrace` command line tool bundled with Xcode.

While the profiler is tested and seems to be working well, I consider it rather a preliminary version and open to a discussion on what and how it should measure.

Currently, the profiler only supports PMU counters sampling using `CPU Counters` instrument provided by the Instruments app / xctrace.
Unfortunately, `CPU Counters` instrument has no default settings, unlike `Time Profiler` and `CPU Profiler` instruments used by the recently merged `xctraceasm` profiler.
To use `CPU Counters`, a user has to create a template in the Instruments UI, select PMU events, save the template and then supply to `xctracenorm` as an argument.

This workflow not only prevents use of the profiler without preliminary manual configuration, but also tends to be annoying when it comes to measuring multiple events, as xctrace, unlike perf_events, does not support events multiplexing.

Thankfully, command-line-based configuration and default parameters could be emulated by building a custom Instruments package that imports data from `CPU Counters` and also supplies all required parameters.
As you can guess, there's no way to get information about supported PMU events directly from xctrace, but it could be fetched from KPEP database files, stored is `/usr/share/kpep`.
`xctracenorm` relies on that data to validate events specified by a user, if any, and also to print a help message that gives some insights into what could be sampled.

To sum up, there are a few things that were implemented to make `xctracenorm` profiler works:
- CPU model deletion using `sysctl`;
- KPEP file parsing to extract information about the PMU and all supported events;
- selected performance events validation;
- Instruments package building (generate XML, call a builder tool), packages are cached in `~/Library/Caches/org.openjdk.jmh`;
- xctrace execution, resulting samples extraction, and aggregation;
- samples postprocessing to calculate some additional metrics, like CPI and branch missprediction ratio.

Currently, if a user didn't specify any additional options, `xctracenorm` will sample instructions, cycles, branches and mispredicted branches events. 
These were selected as events that should be supported in all hardware macOS runs on; only 4 events were selected for the same reason.

Profiling results look like this on M2-based MacBook:
```
java -jar ./benchmarks.jar -prof xctracenorm -f 1 JMHSample_35_Profilers.Atomic
...
Benchmark                                                                 Mode  Cnt   Score   Error                               Units
JMHSample_35_Profilers.Atomic.test                                        avgt    5   4.055 ± 0.185                               ns/op
JMHSample_35_Profilers.Atomic.test:BRANCH_MISPRED_NONSPEC                 avgt       ≈ 10⁻⁵                                        #/op
JMHSample_35_Profilers.Atomic.test:Branch miss ratio                      avgt       ≈ 10⁻⁶          BRANCH_MISPRED_NONSPEC/INST_BRANCH
JMHSample_35_Profilers.Atomic.test:CORE_ACTIVE_CYCLE                      avgt       10.541                                        #/op
JMHSample_35_Profilers.Atomic.test:CPI                                    avgt        0.351                  CORE_ACTIVE_CYCLE/INST_ALL
JMHSample_35_Profilers.Atomic.test:INST_ALL                               avgt       30.031                                        #/op
JMHSample_35_Profilers.Atomic.test:INST_BRANCH                            avgt        3.850                                        #/op
JMHSample_35_Profilers.Atomic.test:INST_BRANCH density (of instructions)  avgt        0.128                        INST_BRANCH/INST_ALL
JMHSample_35_Profilers.Atomic.test:IPC                                    avgt        2.849                  INST_ALL/CORE_ACTIVE_CYCLE

```

Here are some alternatives to existing implementation (of default parameters mode, mostly):
- Create an Instruments template and bundle it with JMH instead of generating a package dynamically: templates have a proprietary and obscure format, it's unclear if a template created on one device will work on other devices, or whether it continues working well after Xcode update. Also, dynamic package generation facilitates PMU events selection in CLI when running benchmarks, which seems to be much more convenient that setting up templates in UI.
- Don't parse KPEP files: data from these files used for validation, and also allows to gather info about what could be sampled without going to a separate tool. Validation is necessary as xctrace simply crashes when something is wrong with selected events.

Open questions and things that require some fixes:
* [x] Currently, only PMU sampling was supported, however, profiler may also sample some software events like context switches, virtual memory operations and syscall statistics. It's unclear if all that should ever be supported.
* [x] Some parts of the profiling process, namely KPEP parsing and Instruments package building are covered by tests in `jmh-core-it` module (`XCTraceSupportTest`). Tests functions are package-private in `jmh-core`, so to test them on `it` module I had to call everything through reflection, which definitely doesn't look good. I'm not sure how to both keep the API private and test in another module, so I'm looking forward to any advice on that.
* [x] `CPU Counters` doesn't work inside VMs (at least, because there are no kpep-files for VM CPU ids, so xctrace can't load those files to fetch info about PMU events; and even if a KPEP file is there, there's no way to sample PMU counters inside VM), so all positive profiler's tests are currently skipped inside GH agents. I'm not sure what could be done here to improve testability. 
* [x] When running tests locally, surefire forks used in `jmh-core-it` cause interference between `xctraceasm` and `xctracenorm` tests leading to test failures. I'm currently looking into how to overcome the issue and serialize these tests' execution. In the worst case, tests for both profilers could be placed in a single test class, I guess.

Speaking of testing, I manually ran basic scenarios (listing profilers, printing a help message, listing supported events, and, finally, profiling) on Intel-, M1- and M2-based MacBooks.
Results and a script I used to collect data could be found here: https://gist.github.com/fzhinkin/f7c5db00f5e3417191d66994ed880818

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [CODETOOLS-7903722](https://bugs.openjdk.org/browse/CODETOOLS-7903722): JMH: Add xctrace-based perfnorm profiler for macOS (**Enhancement** - P4)

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jmh.git pull/131/head:pull/131` \
`$ git checkout pull/131`

Update a local copy of the PR: \
`$ git checkout pull/131` \
`$ git pull https://git.openjdk.org/jmh.git pull/131/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 131`

View PR using the GUI difftool: \
`$ git pr show -t 131`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jmh/pull/131.diff">https://git.openjdk.org/jmh/pull/131.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jmh/pull/131#issuecomment-2097517384)

### CODE REVIEW

**Comment by fzhinkin** [comment]
> Here are some alternatives to existing implementation (of default parameters mode, mostly):
> ...
> - Don't parse KPEP files: data from these files used for validation, and also allows to gather info about what could be sampled without going to a separate tool. Validation is necessary as xctrace simply crashes when something is wrong with selected events.

On the other hand, KPEP parsing and PMU events validation logic is a fair share of the whole PR, thus removing it should make the implementation much simpler.

As a middle ground, the validation could be reduced to checking event names only (and that could be done without parsing KPEP XML file, by running `plutil -extract system.cpu.events raw -o - <path.to.plist>`).

@shipilev please let me know which option (if any at all) suits JMH better.

**Comment by fzhinkin** [comment]
Gentle ping to preserve the PR from being closed.

**Comment by galderz** [comment]
Great work @fzhinkin! I've been trying this out and thought I'd share my experience:

* `xctraceasm` seems to be working well, but some characters do not appear to be printed well:

```
   0.03%  ??? ??? ?   0x00000001168cf5d8:   sbfiz		x10, x6, #2, #0x20  ;*iaload {reexecute=0 rethrow=0 return_oop=0}
```

The `?` characters should instead be the vertical lines that bind together loops.

* `xctracenorm` not working for me, it stops with error:

```
Can't create an Instruments package:

Finished running test ...
```

I do have `instrumentbuilder` locally but any calls to it get killed:

```
$ ls -l /Applications/Xcode.app/Contents/Developer/usr/bin/instrumentbuilder
-rwxr-xr-x  1 galder  staff   184K  6 Oct  2023 /Applications/Xcode.app/Contents/Developer/usr/bin/instrumentbuilder

$ /Applications/Xcode.app/Contents/Developer/usr/bin/instrumentbuilder -h
[1]    98880 killed     /Applications/Xcode.app/Contents/Developer/usr/bin/instrumentbuilder -h
```

I'm on Sonoma 14.5.

**Comment by galderz** [comment]
The issue with `instrumentbuilder` was due to an old xcode version. Updated xcode and then it worked fine.

**Comment by fzhinkin** [comment]
@shipilev, thank you for taking a look and addressing the questions!

> So I think we should be doing a favor to ourselves here, and only implement a very basic support: no building packages, no selectable events, a preconfigured template (which we might later conditionalize on xcode version?) with most useful events, an ability to substitute your own template (for power users). Yes, it would not be as great as perfnorm, but that's the tradeoff we make not to deal with the profiling interface clunkiness.
> 
> Does that work?

Absolutely! I'll come back by with a trimmed down version soon.

**Comment by fzhinkin** [comment]
I updated the profiler so it no longer generates Instruments packages. Instead, users need to specify Instruments template name or path. 
If template was not specified, a default template will be used instead. It samples four PMU counters which, through aliased names, should be available on all supported CPUs.

I tested the profiler with Instruments versions from `13.0` to `16.0`. The default template works fine with any of them, on both x86-64 and aarch64 devices.

**Review by shipilev — COMMENTED** [review]
Nice!

I tried on my Mac, and I think profiler requires super-user privileges? I only succeeded when running `sudo java ...`. Some JMH profilers supply sudo automatically, this should do the same?

I also have a comment about units:

```
Benchmark                                                                 Mode  Cnt   Score   Error                               Units
JMHSample_35_Profilers.Atomic.test                                        avgt    5   4.055 ± 0.185                               ns/op
JMHSample_35_Profilers.Atomic.test:BRANCH_MISPRED_NONSPEC                 avgt       ≈ 10⁻⁵                                        #/op
JMHSample_35_Profilers.Atomic.test:Branch miss ratio                      avgt       ≈ 10⁻⁶          BRANCH_MISPRED_NONSPEC/INST_BRANCH
JMHSample_35_Profilers.Atomic.test:CORE_ACTIVE_CYCLE                      avgt       10.541                                        #/op
JMHSample_35_Profilers.Atomic.test:CPI                                    avgt        0.351                  CORE_ACTIVE_CYCLE/INST_ALL
JMHSample_35_Profilers.Atomic.test:INST_ALL                               avgt       30.031                                        #/op
JMHSample_35_Profilers.Atomic.test:INST_BRANCH                            avgt        3.850                                        #/op
JMHSample_35_Profilers.Atomic.test:INST_BRANCH density (of instructions)  avgt        0.128                        INST_BRANCH/INST_ALL
JMHSample_35_Profilers.Atomic.test:IPC                                    avgt        2.849                  INST_ALL/CORE_ACTIVE_CYCLE
```

The large `CORE_ACTIVE_CYCLE/INST_ALL` are not great: too long? See how Linux perfnorm reports them, e.g. `clks/insns`.

Other code nits:

File: jmh-core-it/src/test/java/org/openjdk/jmh/it/profilers/XCTraceNormProfilerTest.java [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+                IllegalStateException.class, () -> new Runner(opts).runSingle());
+    }
+
+    @Test
+    public void testUseCustomTemplate() throws Exception {
+        skipIfProfilerNotSupport();
+        skipIfRunningInsideVirtualMachine();
+
+        RunResult result;
+        File templateFile = FileUtils.extractFromResource("/XCTraceNormTestTemplate.xml");
```
[the remark this answers is not in the document]
Comment (fzhinkin) [reply jmh-core-it/src/test/java/org/openjdk/jmh/it/profilers/XCTraceNormProfilerTest.java:171]: It uses different event settings. Given the test checking that not-existing / wrong template won't work, it indeed should be fine to result the default template.

**Comment by fzhinkin** [comment]
@shipilev 
> I tried on my Mac, and I think profiler requires super-user privileges? I only succeeded when running `sudo java ....`

It should not require super-user privileges. Otherwise, `xctraceasm` will require them too.
Could you please elaborate on what happened w/o SU-privileges?

**Comment by fzhinkin** [comment]
> The large `CORE_ACTIVE_CYCLE/INST_ALL` are not great: too long? See how Linux perfnorm reports them, e.g. `clks/insns`.

I'll use aliases similar to what `perf_events` provides, then.

**Comment by shipilev** [comment]
> > I tried on my Mac, and I think profiler requires super-user privileges? I only succeeded when running `sudo java ....`
> 
> It should not require super-user privileges. Otherwise, `xctraceasm` will require them too. Could you please elaborate on what happened w/o SU-privileges?

Yeah, `xctraceasm` does not work either. This might be a borked setup on my side, really. 

```
$ java -jar jmh-samples/target/benchmarks.jar -prof xctracenorm -f 1 JMHSample_35_Profilers.Atomic -v EXTRA
...

# Fork: 1 of 1
# Preparing profilers: XCTraceNormProfiler 
[-[SimServiceContext sendRequest:]:1764] ERROR : Unable to deliver request ({
    "developer_dir" = "/Applications/Xcode.app/Contents/Developer";
    request = "set_developer_dir";
}) because we are not connected to CoreSimulatorService.
[-[SimServiceContext initWithDeveloperDir:connectionType:error:]:450] WARN  : Unable to discover any Simulator runtimes. Developer Directory is /Applications/Xcode.app/Contents/Developer.
[-[SimDiskImageManager init]:146] ERROR : Could not kickstart simdiskimaged; SimDiskImageManager services will not be available: Error Domain=NSPOSIXErrorDomain Code=53 "Software caused connection abort" UserInfo={NSLocalizedDescription=Error returned in reply from CoreSimulatorService: Connection invalid}
[-[SimServiceContext sendRequest:reply:error:]:1784] ERROR : Unable to deliver request ({
    request = "notification_subscription";
    "set_path" = "/Users/shipilev/Library/Developer/CoreSimulator/Devices";
}) because we are not connected to CoreSimulatorService.
Starting recording with the jmh14925377671874505903extracted template. Launching process: java. 
xpc connection invalid: com.apple.dt.instruments.dtsecurity.xpc
```

File: jmh-core/src/main/java/org/openjdk/jmh/profile/XCTraceNormProfiler.java [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+
+    private static XCTraceTableHandler.XCTraceTableDesc findTableDescription(XCTraceTableOfContentsHandler tocHandler) {
+        XCTraceTableHandler.XCTraceTableDesc tableDesc = tocHandler.getSupportedTables()
+                .stream()
+                .filter(t -> t.getTableType() == SUPPORTED_TABLE_TYPE)
+                .findFirst()
+                .orElseThrow(() -> new IllegalStateException("Table \"" + SUPPORTED_TABLE_TYPE.tableName +
+                        "\" was not found in the trace results."));
+        if (tableDesc.getPmcEvents().isEmpty() && tableDesc.getTriggerType() == XCTraceTableHandler.TriggerType.TIME) {
+            throw new IllegalStateException("Results does not contain any events.");
```
[the remark this answers is not in the document]
Comment (fzhinkin) [reply jmh-core/src/main/java/org/openjdk/jmh/profile/XCTraceNormProfiler.java:153]: Effectively, it'll be thrown from `afterTrial`, and `afterTrial` has no exceptions in its signature, so I'm not sure if it's worth throwing `ProfilerException` only to catch it, wrap it into some `RuntimeException`-subclass and rethrow it.

**Comment by fzhinkin** [comment]
Abbreviated units, not it looks like:
```
Benchmark                                                          Mode  Cnt   Score   Error      Units
JMHSample_35_Profilers.Atomic.test                                 avgt    5   1.885 ± 0.006      ns/op
JMHSample_35_Profilers.Atomic.test:BRANCH_MISPRED_NONSPEC          avgt       ≈ 10⁻⁵               #/op
JMHSample_35_Profilers.Atomic.test:Branch miss ratio               avgt       ≈ 10⁻⁵           miss/brs
JMHSample_35_Profilers.Atomic.test:CORE_ACTIVE_CYCLE               avgt        6.988               #/op
JMHSample_35_Profilers.Atomic.test:CPI                             avgt        0.584          clks/insn
JMHSample_35_Profilers.Atomic.test:INST_ALL                        avgt       11.970               #/op
JMHSample_35_Profilers.Atomic.test:INST_BRANCH                     avgt        1.995               #/op
JMHSample_35_Profilers.Atomic.test:INST_BRANCH density (of insns)  avgt        0.167             #/insn
JMHSample_35_Profilers.Atomic.test:IPC                             avgt        1.713          insns/clk
```

**Comment by fzhinkin** [comment]
> Yeah, `xctraceasm` does not work either. This might be a borked setup on my side, really.

I had a similar issue when a custom kernel trace client started w/ SU-privileges and died without releasing `ktrace`. Long story short, rebooting the device helped (~~`xctrace` start running, but my system-programming abilities didn't improve~~).

Messages about device simulator looks suspicious, maybe `xctrace list devices` will shed some light (perhaps, there are no devices available, due to `ktrace` being locked by another process, so xctrace decided to run on an emulator?).

**Review by shipilev — APPROVED** [review]
I think this is good for MVP :)

**Comment by fzhinkin** [comment]
Thank you, @shipilev!

**Pull request closed without merging**